### ISSUE TITLE
Navigating from router-view throws Missing required param

### ISSUE THREAD

**Issue by phillduffy** [vuejs/vue-router-next#845]
### Version
4.0.5

### Reproduction link
[https://github.com/phillduffy/route-issue](https://github.com/phillduffy/route-issue)

### Steps to reproduce
- Run the repo
- Navigate to 'Child with Param'
- Click 'Back Home'

### What is expected?
User should navigate back to the home page

### What is actually happening?
Exception is being thrown

`runtime-core.esm-bundler.js:38 [Vue warn]: Unhandled error during execution of watcher callback 
  at <RouterLink to= {name: "grandchild"} > 
  at <Child onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< Proxy {…} > > 
  at <RouterView> 
  at <Home onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< Proxy {…} > > 
  at <RouterView> 
  at <App>`

and

`Uncaught (in promise) Error: Missing required param "id"
    at Object.stringify (vue-router.esm-bundler.js:994)
    at Object.resolve (vue-router.esm-bundler.js:1410)
    at Object.resolve (vue-router.esm-bundler.js:2860)
    at vue-router.esm-bundler.js:2039
    at ComputedRefImpl.reactiveEffect [as effect] (reactivity.esm-bundler.js:42)
    at ComputedRefImpl.get value [as value] (reactivity.esm-bundler.js:828)
    at vue-router.esm-bundler.js:2041
    at ComputedRefImpl.reactiveEffect [as effect] (reactivity.esm-bundler.js:42)
    at ComputedRefImpl.get value [as value] (reactivity.esm-bundler.js:828)
    at vue-router.esm-bundler.js:2064`

---
Updating the grandchild link to the following gets around the issue

`<router-link :to="{ name: 'grandchild', params: { id: 1 } }"
          >Grandchild</router-link
        >`

**Comment by posva** [comment]
Good news is the bug disappears in production because the watchers are only there for devtools

**Comment by alijuniorbr** [comment]
For those who need and are interested in fixing this problem before the new release of vue-router v4.0.6 can do the following:

in your project directory:

remove the vue-router:
`npm r vue-router`

install the current version (already corrected, but not published) of the vue-router:
`npm i https://github.com/vuejs/vue-router-next/tarball/master`

install dependencies and build vue-router:
```
cd node_modules\vue-router
npm i
npm run build
```
that's all! now your project contains these corrections that have already been committed, but are awaiting release v4.0.6

**of course, don't use this build in production, because it can contain some bugs that will be prevented from v4.0.6 release**

**Comment by tgel0** [comment]
For me, the error was caused by a route guard where I didn't add the required param actually. Maybe someone ends up reading this who made the same mistake as me.

**Comment by KaKi87** [comment]
Hello,

I was experiencing the very same issue when using `this.$route.params` and `this.$router.resolve` in `computed` properties, although I can't seem to reproduce it on JSFiddle.

My guess is, values inside `params` turn into `undefined` between the time when the navigation starts and the time when the component is unmounted.

Because, checking that the `params` object wasn't empty is the only small fix I needed to prevent the error from being thrown : https://git.kaki87.net/KaKi87/scraper-instagram-gui/commit/85b0377d5bda2108926fa714b94642317c43c9f2

Thanks

**Comment by pwang2** [comment]
Can we reopen this? I don't think the `has workaround` works for the others suffering this especially in a migration project from vue2.   :(

**Comment by villageseo** [comment]
Still actual problem in "vue-router": "4.1.6".

**Comment by villageseo** [comment]
Maybe it will help someone, but specially in my case the problem was in vuetify's v-tab component with missing params in **to** attribute. Damn it's not obvious. 
 
So when I've added: **params: $route.params** -> it helped. 

Full example: 
` <v-tab
      v-for="(item, key) in items"
      :key="key"
      :disabled="item.disabled"
      :to="plain ? null : { name: item.name, params: $route.params }" // <- here
    >
     <span>{{ $t(item.humanName) }}</span>
    </v-tab>`

**Comment by MatthewAry** [comment]
> Still actual problem in "vue-router": "4.1.6". (in development mode only)

I agree. I have been able to reproduce this issue with `RouterLink` but I have not been able to create a sharable reproduction, and it's not been without trying! I can get this issue to show up in my internal app but when I try to recreate the same conditions in a simplified reproduction I can't get the error to present. My learnings so far:
- This issue presents itself in prod
- This issue affects vuetify components that create router links (See https://github.com/vuetifyjs/vuetify/issues/17176)
- This issue does not seem to be related to any devtool code as far as I can tell

I worry that I suck at explaining thing which is why I wrote what's below
- Basically I would expect that you don't need to put all params in a `RouteLocationNamedRaw` object if they can be inherited by the current route.
- I don't want to have to pass in params for a router link when that particular router link appears on a route already has the param.
- The link is on a component in a child route, and the param is already defined on the parent route, so if the RouterLink on the child is for another route on the same parent, then it should inherit the param.

**Comment by Oleksii14** [comment]
In my case, the reason was in [this change](https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22).

So I changed all `router.push` & `router.replace` accordingly, it fixed my issue.

Example:

```ts
createRouter({
  routes: [
    {
      path: "/:dynamicVariable",
      name: "parent",
      redirect: { name: "child" },
      children: [
        {
          name: "child",
          path: "child"
        }
      ]
    },
  ]
})
```

Causes error:

```ts
$router.push({ name: "child", params: { dynamicVariable: "1" } })
```

Resolution:

```ts
$router.push({ name: "parent", params: { dynamicVariable: "1" } })
```

**Comment by Oleksii14** [comment]
@MatthewAry that is right. I am also experiencing these issues with Vuetify. The temporary solution is not to use `v-btn` as router link:

```
<v-btn :to="route" />
```
Change to
```
<v-btn @click="$router.push(route)" />
```

**Comment by MatthewAry** [comment]
Please see my comment on what I have learned so far. https://github.com/vuetifyjs/vuetify/issues/17176#issuecomment-1517843304

**Comment by postmodernistx** [comment]
My five cents:

I had the same error, `Error: Missing required param "id"`, when navigating back to the "top level route" from a second level subpage. This was happening in one project, but not in the other.

Example navigation: `Home/index page (OK) → Some mid-level page (OK) → Second level page (OK) → Back to index (error)`

In the erroneous project, I had "top-level dynamic routes", whereas in the working one, each "top level path" was prefixed with something before entering dynamic routes.

Perhaps not a feasible solution, but at least it got rid of the error for now.

Example, not working:

```vue
routes: [
  {
    path: '/',
    name: 'Page0',
    component: () => import('../views/Page0.vue'),
    children: [
      {
        name: 'HomeSubpage',
        path: ':homedynamic',
        component: () => import('../views/HomeSubpage.vue'),
      },
    ],
  },
  {
    path: '/:dynamic', // Issue here 🚨
    name: 'Page1',
    component: () => import('../views/Page1.vue'),
    children: [
      {
        path: ':subpage',
        name: 'Page2',
        component: () => import('../views/Page2.vue'),
      },
    ],
  },
```

And working version:

```vue
routes: [
  {
    path: '/',
    name: 'Page0',
    component: () => import('../views/Page0.vue'),
    children: [
      {
        name: 'HomeSubpage',
        path: ':dynamic',
        component: () => import('../views/HomeSubpage.vue'),
      },
    ],
  },
  {
    path: '/expenses', // Working when not a param ✅
    name: 'Page1',
    component: () => import('../views/Page1.vue'),
    children: [
      {
        path: ':subpage',
        name: 'Page2',
        component: () => import('../views/Page2.vue'),
      },
    ],
  },
```

I suppose I could make the param optional as well, as so (and have an in-component check instead for its existence):
```
path: '/:dynamic?',
```

Looking at the docs, I see no code samples with "top-level dynamic names", so maybe not supported. So implementation of navigation failure guards is probably the best way to go ;)

https://router.vuejs.org/guide/advanced/navigation-failures.html#Global-navigation-failures

And in the original sample code, as the "subroute" is prefixed with a beginning `/`, it becomes a top-level dynamic route? https://github.com/phillduffy/route-issue/blob/main/src/routes.js#L10

**Comment by acnowland** [comment]
Hmm, its interesting that it happens so many different ways tbh. I also feel like this is just getting completely ignored though by the router devs due to the labeled workaround, but its clearly still occuring. We are mainly noticing it when we are having users hit the back button. Have tried switching things from being computed properties, tried using postFlush, upgraded the router, but continues to occur. No way to actually make a replication of how we use our app but its a very large app and uses alot of dynamic params that are passed around.

**Comment by kenneth4896** [comment]
Still actual problem in "vue-router": "4.1.6". (in development mode only)

**Comment by MalindiJohn** [comment]
Still having the same issue in vue-router 4.1.6

```
vue-router.js?v=f9d707fe:552 Uncaught (in promise) Error: No match for
 {"name":"index","params":{}}
    at createRouterError (vue-router.js?v=f9d707fe:552:19)
    at Object.resolve (vue-router.js?v=f9d707fe:1004:15)
    at Object.resolve (vue-router.js?v=f9d707fe:2160:34)
    at resolveNavLinkRouteName (utils.ts:38:3)
    at Proxy.isNavLinkActive (utils.ts:50:3)
    at Proxy._sfc_render (HorizontalNavLink.vue:33:44)
    at renderComponentRoot (chunk-RMIPITSQ.js?v=f9d707fe:2250:39)
    at ReactiveEffect.componentUpdateFn [as fn] (chunk-RMIPITSQ.js?v=f9d707fe:5516:46)
    at ReactiveEffect.run (chunk-RMIPITSQ.js?v=f9d707fe:1418:23)
    at instance.update (chunk-RMIPITSQ.js?v=f9d707fe:5615:52)
```

Any workaround here?

**Comment by rwam** [comment]
Having the same issue using vue-router@4.2.5. I can solve it making all dynamic params optional but a more stable solution would be preferred.

**Comment by dante4rt** [comment]
Having same issue in vue-router 4.0.0.

**Comment by Bekzod101001** [comment]
Having same issue in vue-router 4.2.5, when using with vuetify navigation drawer

**Comment by bbedyukh** [comment]
Having same issue in vue-router 4.2.5.

**Comment by JhumanJ** [comment]
Same issue in latest version of Nuxt

**Comment by 1nusah** [comment]
hi, is there any fix for this? major bug with my nuxt app

**Comment by shashankgaurav17** [comment]
still happening..not fixed

**Comment by 1nusah** [comment]
yes, make your params for nested routes optional. that's the work around i found

**Comment by jedifunk** [comment]
@1nusah this does not always work.

```{
        path: 'home/:yearParam',
        name: 'Year',
        component: () => import('@/views/YearView.vue'),
        props: true,
      },
      {
        path: 'home/:yearParam/:dateParam',
        name: 'Date',
        component: () => import('@/views/ShowView.vue'),
        props: true,
      },```
      
      adding '?' to `/:yearParam` errors the parent, adding ? to `/:yearParam?/:dateParam` makes parent and then year work, but /year/date errors

**Comment by edimitchel** [comment]
I still have this issue with Nuxt and nested route. My params is optional and set as the same level than a sibling route without param. The only way for me is to use route name instead of route pass.

**Comment by Kilian-Pichard** [comment]
I confirm edimitchel, I still have the same issue with Nuxt 3.11.2 ans vue-router 4.3.2.
The two following components doesn't work: 
`<NuxtLink :to="`/id`">id</NuxtLink>`
`<NuxtLink :to="{ name: 'id', params: { id: 'id' } }">Username</NuxtLink>`

**Comment by edimitchel** [comment]
Let's build a reproduction to help the resolution of this issue.

I made it work by using Vue router directly instead of using `navigateTo` method.

**Comment by Kilian-Pichard** [comment]
@edimitchel, I'm in! How do you want to do it?

And how did you get it to work using the Vue router directly?

**Comment by edimitchel** [comment]
I start a Stackblitz project: https://stackblitz.com/edit/nuxt-starter-4crj1a?file=app%2Fpages%2Findex.vue
(it uses the next directory organisation coming from upcoming Nuxt 4 version)

I had a side effect issue about this error where when I prevent a redirection by using `onBeforeRouteLeave` then doing the `navigateTo` on the root of the app, nothing happen, using `useRouter().push` method fixed the problem.

**Comment by Kilian-Pichard** [comment]
I forked your Stackblitz project and I tried to replicate my issue. I found it.
The problem I had was that I was calling `route.params.X` from a layout. And apparently layouts don't return routes from `const route = useRoute()`.
Example here: https://stackblitz.com/edit/nuxt-starter-dkd15p?file=app%2Flayouts%2Fdefault.vue

**Comment by JulienDecruydt** [comment]
Hello,

Still have the same bug in NUXT `^3.12.4`.
When clicking to go back to landing page, the issue is throw even if `landing.vue` doesn't need any param.

`Uncaught (in promise) Error: Missing required param "slug"`

Structure folder :

```
- [slug]
  - index.vue
- landing.vue
```

Component button (nuxt-link behind) : 
```
<CoreButton 
  :to="{
       name: 'landing',
   }" 
   text="Go to home"
/>
```

**Comment by JulienDecruydt** [comment]
Hello,

Still have the same bug in NUXT `^3.12.4`.
When clicking to go back to landing page, the issue is throw even if `landing.vue` doesn't need any param.

Structure folder :

```
- [slug]
  - index.vue
- landing.vue
```

Component button (nuxt-link behind) : 
```
<CoreButton 
  :to="{
       name: 'landing',
   }" 
   text="Go to home"
/>
```

**Comment by edimitchel** [comment]
@JulienDecruydt do you use a component library? The issue I got was created by a bad usage of NuxtLink from a Vuetify component. By using native NuxtLink instead of TabLink, I resolve my issue.

**Comment by JulienDecruydt** [comment]
> @JulienDecruydt do you use a component library? The issue I got was created by a bad usage of NuxtLink from a Vuetify component. By using native NuxtLink instead of TabLink, I resolve my issue.

Thanks for your help ! But I don't use a components library.

I've solved the issue. 
If someone has the same problem and others solutions don't work, try this : 

- Add `<Suspense>` around your `NuxtLayout` and `NuxtPage`
- Add `<Suspense>` around your `<slot>`

SSR doesn't hydrate params of `<nuxt-link>` and nuxt considerate this has empty when it's rendered

Reference : [here](https://stackoverflow.com/questions/69711253/error-missing-required-param-id-vue-router-router-link)

**Comment by tobychidi** [comment]
I have been having this issue as well. Why is it closed?

**Comment by kevinkosterr** [comment]
Also experiencing this issue (`vue-router@4.5.0`). Like others have mentioned, why is this closed? I found [this article](https://pajuhaan.medium.com/fixing-vue-3-router-missing-params-issue-especially-when-using-vuetify-ed174bcabff7) to be helpful in providing a temporary solution.

**Comment by rissole** [comment]
I managed to stop this error from occurring in dev in my project by commenting out the `watchEffect` that is included in `useLink` when devtools are enabled.

This likely removes or breaks some dev tools functionality -- from what I can tell, it just removes some labels in the Components Tree view that point out which routes are active. If you rely on this feature then this change might not be suitable for you.

If you use [patch-package](https://github.com/ds300/patch-package) you can apply this change in your project by

1. Creating a new file **patches/vue-router+4.5.0.patch** with the following content:

```diff
diff --git a/node_modules/vue-router/dist/vue-router.mjs b/node_modules/vue-router/dist/vue-router.mjs
index 399f57c..d723b9d 100644
--- a/node_modules/vue-router/dist/vue-router.mjs
+++ b/node_modules/vue-router/dist/vue-router.mjs
@@ -2334,14 +2334,14 @@ function useLink(props) {
             instance.__vrl_devtools = instance.__vrl_devtools || [];
             // @ts-expect-error: this is internal
             instance.__vrl_devtools.push(linkContextDevtools);
-            watchEffect(() => {
-                linkContextDevtools.route = route.value;
-                linkContextDevtools.isActive = isActive.value;
-                linkContextDevtools.isExactActive = isExactActive.value;
-                linkContextDevtools.error = isRouteLocation(unref(props.to))
-                    ? null
-                    : 'Invalid "to" value';
-            }, { flush: 'post' });
+            // watchEffect(() => {
+            //     linkContextDevtools.route = route.value;
+            //     linkContextDevtools.isActive = isActive.value;
+            //     linkContextDevtools.isExactActive = isExactActive.value;
+            //     linkContextDevtools.error = isRouteLocation(unref(props.to))
+            //         ? null
+            //         : 'Invalid "to" value';
+            // }, { flush: 'post' });
         }
     }
     /**
```
